### PR TITLE
[ao][ns] PNP demo for exposing arbitrary model transforms

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89892

Adding way to use arbitrary prepare and convert functions with PNP.

Differential Revision: [D41167110](https://our.internmc.facebook.com/intern/diff/D41167110/)

### CODE REVIEW

File: torch/ao/ns/_numeric_suite_fx.py [hunk 1]
```diff
@@ -752,6 +752,9 @@ def prepare_n_shadows_model(
     example_inputs: Any,
     qconfig_multi_mapping: QConfigMultiMapping,
     backend_config: BackendConfig,
+    exposed_prepare_function: Any=None,
```
Comment (vkuzo) [inline torch/ao/ns/_numeric_suite_fx.py:755]: nit: fix lint

Comment (vkuzo) [reply torch/ao/ns/_numeric_suite_fx.py:755]: would a name such as `prepare_fn_override` and `convert_fn_override`, or `custom_prepare_fn` / `custom_convert_fn` make more sense?  the `exposed` word seems ambiguous

File: torch/ao/ns/_numeric_suite_fx.py [hunk 2]
```diff
@@ -862,7 +870,7 @@ def loggers_set_save_activations(
         if isinstance(child, OutputLogger):
             child.save_activations = save_activations
 
-def convert_n_shadows_model(model: GraphModule) -> GraphModule:
+def convert_n_shadows_model(model: GraphModule, exposed_convert_function: Any=None, exposed_convert_kwargs: Any=None) -> GraphModule:
```
Comment (vkuzo) [inline torch/ao/ns/_numeric_suite_fx.py:873]: can we use `Callable` and `Dict[str, Any]` as the respective types?

File: torch/ao/ns/_numeric_suite_fx.py (hunk 1 above)
Comment (HDCharles) [reply torch/ao/ns/_numeric_suite_fx.py:755]: ok, will update

File: torch/ao/ns/_numeric_suite_fx.py (hunk 2 above)
Comment (HDCharles) [reply torch/ao/ns/_numeric_suite_fx.py]: fixed

File: torch/ao/ns/fx/n_shadows_utils.py [hunk 3]
```diff
@@ -566,9 +568,19 @@ def handle_subgraph_candidate(
...
-            orig_mod_copy_wrapped, qconfig_mapping, example_inputs=example_inputs,
-            prepare_custom_config=prepare_custom_config)
+        if custom_prepare_fn is None:
+            orig_mod_copy_wrapped = torch.ao.quantization.quantize_fx.prepare_fx(
+                orig_mod_copy_wrapped, qconfig_mapping, example_inputs=example_inputs,
+                prepare_custom_config=prepare_custom_config)
+        else:
+            if custom_prepare_kwargs is None:
+                custom_prepare_kwargs = {}
+            custom_prepare_kwargs["example_inputs"] = example_inputs
```
Comment (vkuzo) [inline torch/ao/ns/fx/n_shadows_utils.py:578]: should we check that the original `custom_prepare_kwargs` don't specify `example_inputs`, `prepare_custom_config`, `qconfig_mapping`?

**Review by vkuzo — APPROVED** [review]

File: torch/ao/ns/fx/n_shadows_utils.py (hunk 3 above)
Comment (HDCharles) [reply torch/ao/ns/fx/n_shadows_utils.py:578]: yeah, that makes sense

**Pull request closed without merging**